src/v/kafka/client/client.cc

```diff
@@ -541,21 +541,26 @@ ss::future<kafka::fetch_response> client::consumer_fetch(
                 std::chrono::duration_cast<std::chrono::milliseconds>(timeout),
                 max_bytes);
           })
-          .then([this](kafka::fetch_response res) {
-              bool has_error = std::any_of(
-                res.data.topics.begin(),
-                res.data.topics.end(),
-                [](auto const& topics) {
-                    return std::any_of(
-                      topics.partitions.begin(),
-                      topics.partitions.end(),
-                      [](const auto& p) {
-                          return p.error_code != error_code::none;
-                      });
-                });
-              return (has_error ? _wait_or_start_update_metadata() : ss::now())
-                .then(
-                  [res{std::move(res)}]() mutable { return std::move(res); });
+          .then([](kafka::fetch_response res) {
+              for (const auto& topic : res.data.topics) {
+                  for (const auto& partition : topic.partitions) {
+                      if (partition.error_code != error_code::none) {
+                          vlog(
+                            kclog.info,
+                            "AAA has error {}",
+                            partition.error_code);
+                          model::topic_partition tp{
+                            topic.name, partition.partition_index};
+                          // Mitigate error will force a metadata refresh on
+                          // several error types such as
+                          // not_leader_for_partition, so the request should
+                          // eventually succeed with retries.
+                          return ss::make_exception_future<fetch_response>(
+                            partition_error(tp, partition.error_code));
```

Comment:
Couldn't we modify the existing code to be something like:
```
if (has_error) {
  return ss::make_exception_future<fetch_response>(partition_error(<first tp>, <first errc>);
}
```
so your saying the fact that we throw here, goes down a path that will grab metadata AND retry the fetch request? And the previous code was only updating metadata and NOT retrying?

Reply:
> Couldn't we modify the existing code to be something like:
> 
> ```
> if (has_error) {
>   return ss::make_exception_future<fetch_response>(partition_error(<first tp>, <first errc>);
> }
> ```

Yes that would work. I didn't want to add the extra control flow though, but perhaps I was too hasty. I'll make the change.

> so your saying the fact that we throw here, goes down a path that will grab metadata AND retry the fetch request? 

Correct.

> And the previous code was only updating metadata and NOT retrying

Correct

Now that you write it out, perhaps the solution is to **not** retry within our kafka client but instead expect the caller to retry (i.e., re-issue the REST request).

Reply:
Cool thanks for clarifying that

I think @BenPope would have more insight here, but my initial thoughts would be that the original intention wasn't for retry-able errors to be propagated back to the client for retrying.

Reply:
The intention was to return the results that had been fetched, but before doing so, update metadata, so that the next fetch will hopefully know where the partition leader is.